### PR TITLE
ghub-wait: Generalize for forges other than github

### PR DESCRIPTION
**Pull request by stepnem**
(Immediate motive being adding gitlab.com forking support to forge.el.)

### CODE REVIEW

**Comment by tarsius** [comment]
Did you verify that Gitlab really returns immediately like Github does?

**Comment by stepnem** [comment]
> Did you verify that Gitlab really returns immediately like Github does?

I'm not sure I understand "immediately". I thought ghub-wait was meant
to handle the cases when it would _not_ necessarily return immediately?

In any case, it did return "immediately" when I tried... :-D

**Comment by tarsius** [comment]
The problem that `ghub-wait` is trying to solve for Github is as follows: When the client uses the API to create a new repository, then the server responds with 200 OK. However this response does not mean *the requested repository X has been created* it merely means "your request for the creation of repository X has been received, we will try to get that done soon".

(When I used the term "immediately" above, then I meant "as soon as the request has been received the server confirms that it has received the request and intends to act on it" as opposed to "once the requested action has been carried out the server confirms that it is done doing so-". The former takes much less time (i.e. it is "immediate") than the latter.)

That's a problem because the client may wish to modify the repository once it has been created. But doing so is only possible once the repository actually exists. If the client tries to make a change to a repository that has not been fully created yet, then that results in an error. `ghub-wait` does it keeps making requests that check whether the repository has already been created until it has indeed been created or else until it has taken too long and we now consider the creation to have failed.

----

Therefore what I was trying to get you to confirm was whether Gitlab behaves the same way as Github, i.e. as I just have described. It's possible that it does that, but it is also possible that it does not. Ideally the documentation would say how it behaves in this regard. Github's documentation talks about this explicitly. Gitlab's may do so too; please go and check. Otherwise please perform experiments to figure this out. Also try with forking huge repositories not just creating empty new repositories.

**Comment by tarsius** [comment]
Actually please also check what the http status code is. The appropriate one in the case of Github would be "202 Accepted".

**Comment by stepnem** [comment]
Documentation: https://docs.gitlab.com/ee/api/projects.html#fork-project

The request returns "immediately" with HTTP status 201 Created, and the
import_status property "scheduled", which then turns into "started", and
then "finished" after the fork is complete.

The stati are documented here: https://docs.gitlab.com/ee/api/project_import_export.html#import-status

I couldn't think of anything bigger than gitlab-org/gitlab, which took
maybe about a minute to change from "scheduled" to "finished" (guesstimate).

In any case, that doesn't seem any different from GitHub, barred the 201
instead of 202.

File: ghub.el [hunk 1]
```diff
@@ -379,7 +379,8 @@ called after the next request has finished.  Use the function
   (and (assq 'next (ghub-response-link-relations req))
        (or (ghub--retrieve nil req) t)))
 
-(cl-defun ghub-wait (resource &optional duration &key username auth host)
+(cl-defun ghub-wait (resource &optional (duration 64)
+                              &key username auth host (forge 'github))
```
Comment (tarsius) [inline ghub.el]: I think FORGE needs no default and specifying defaults like this is kinda dangerous a caller might provide nil as the value and expect the default to be used. So just stick to setting DURATION as before.

File: ghub.el [hunk 2]
```diff
@@ -389,18 +390,17 @@ waiting as long again as we already waited between all preceding
...
+      (while (not (ghub-request "GET" resource nil
+                                :noerror t
+                                :username username
+                                :auth auth
+                                :host host
+                                :forge forge))
         (message "Waited (%3ss of %ss) for %s..." total duration resource)
         (if (= total duration)
-            (error "Github is taking too long to create %s" resource)
+            (error "%s is taking too long to create %s" forge resource)
```
Comment (tarsius) [inline ghub.el]: Upcase that as before.

**Comment by stepnem** [comment]
>> +(cl-defun ghub-wait (resource &optional (duration 64)
> +                              &key username auth host (forge 'github))
>
> I think FORGE needs no default and specifying defaults like this is
> kinda dangerous a caller might provide nil as the value and expect the
> default to be used. So just stick to setting DURATION as before.

I don't understand the "dangerous" comment, what do you mean? 

Being explicit about the default seemed cleaner to me, as otherwise the
default will leak in the error message below anyway (checking if forge
is nil and defaulting to "Github"), so all we gain by not specifying it
here is potential confusion or harder code maintenance.

>>          (message "Waited (%3ss of %ss) for %s..." total duration resource)
>          (if (= total duration)
> -            (error "Github is taking too long to create %s" resource)
> +            (error "%s is taking too long to create %s" forge resource)
>
> Upcase that as before.

Will do.

**Comment by tarsius** [comment]
> I don't understand the "dangerous" comment, what do you mean?

Someone might end up doing `(ghub-wait "/foo" nil :forge 'gitlab)` and get a `(wrong-type-argument number-or-marker-p nil)`. You might ask "why would anyone do that?"; I don't know but I have done it many times and as a result wish to make such mistakes impossible.

> Being explicit about the default seemed cleaner to me,

The doc-string mentions the default.

**Comment by stepnem** [comment]
> Someone might end up doing `(ghub-wait "/foo" nil :forge 'gitlab)` and
> get a `(wrong-type-argument number-or-marker-p nil)`. You might ask
> "why would anyone do that?"; I don't know but I have done it many
> times and as a result wish to make such mistakes impossible.

I see, thanks. I'll revert that.

> The doc-string mentions the default.

It mentions the timeout, but that is not the issue; I was speaking about
the forge argument -- without the explicit default it looks like we
don't care what it is, but in fact we do, because in concocting the
error message we have to hardcode "Github" if forge is nil, anyway.

**Comment by tarsius** [comment]
> I was speaking about the forge argument -- without the explicit
> default it looks like we don't care what it is, but in fact we do,

We don't document USERNAME, AUTH and HOST either.  It is assumed that user is familiar with `ghub` enough to know that they can look at the documentation of `ghub-request` if they need a refresher on of these arguments.

> because in concocting the error message we have to hardcode "Github"
> if forge is nil, anyway.

That seems... irrelevant.

**Comment by stepnem** [comment]
>> because in concocting the error message we have to hardcode "Github"
>> if forge is nil, anyway.
>
> That seems... irrelevant.

OK.

**Comment by stepnem** [comment]
> Someone might end up doing `(ghub-wait "/foo" nil :forge 'gitlab)` and
> get a `(wrong-type-argument number-or-marker-p nil)`.

BTW, there is another (more serious IMO) gotcha concerning CL
lambda lists:

  (ghub-wait "/foo" :forge 'gitlab)

=> Debugger entered--Lisp error: (error "Keyword argument gitlab not one of (:username :auth :host :forge)")

For this reason mixing &optional and &key is generally considered bad
practice, some CL implementations even give compiler warnings, cf. eg.

http://www.gigamonkeys.com/book/functions.html
https://dept-info.labri.fr/~strandh/Teaching/MTP/Common/David-Lamkins/chapter21.html
https://lispcookbook.github.io/cl-cookbook/functions.html

Wouldn't changing `duration` to a keyword argument (better still, rename
it to :timeout, as that's what it really is) make a better interface?

**Comment by tarsius** [comment]
All the `ghub-METHOD` api functions have the same issue. I think it is to late to change that.

**Comment by stepnem** [comment]
On Fri, 28 Feb 2020 03:38:46 -0800
Jonas Bernoulli wrote:

> All the `ghub-METHOD` api functions have the same issue. I think it is to late to change that.

Hm... epkg only lists 3 packages depending on ghub not authored by you
or Sean. I'd say let's do it now. Having to give nil all the time feels
weird, and the &key error is bound to happen (and the error message is
kind of hard to figure out if you don't know about the issue).

**Comment by tarsius** [comment]
Okay, I'll think about it.

/cc @vermiculus

**Comment by stepnem** [comment]
Thanks.

If needed, I volunteer to do all the work, including pull requests and
support for or communication with affected 3rd parties, or any part of
the above.

**Comment by npostavs** [comment]
> All the `ghub-METHOD` api functions have the same issue. I think it is to late to change that.

Looking at these `ghub-foo` functions, I have to wonder if they're really adding much. They all seem to be very thin wrappers (adding a single param) around `ghub-request`. What if you did

```elisp
(cl-defun ghub (method resource
                       &key params query payload headers
                       silent unpaginate noerror reader
                       username auth host forge
                       callback errorback value extra)
  [ghub-request current body] ...)
```
And then mark all the rest of the `ghub-foo` functions obsolete. The library users would have `(ghub "DELETE" ...)` instead of `(ghub-delete ...)` which seems fine to me.

**Pull request merged**